Feat: Create `RenameGroupForm`

kanta1207 opened the pull request:
<!--

Title Naming Conventions:

- Start with a capitalized prefix followed by an imperative form of a verb, just like a commit message.
- Refer to the "Commit Convention" for the prefix:
  https://github.com/nishiki-tech/nishiki-frontend/blob/develop/docs/CONTRIBUTING.md#commit-convention

Example:
- Fix: Resolve memory leak in data processing module
- Feat: Add two-factor authentication
- Refactor: Improve state management in React components
- Docs: Add Windows setup instructions

-->

## Overview

<!-- Provide a brief description of the changes introduced by this PR. -->
- Create `RenameGroupForm` component, except for api client

## Changes

<!-- List the changes -->
<!-- Delete this section if not needed -->

- Create `GroupCardContent` component separately from `GroupCard` component
- Create `RenameGroupForm` component
- Added `icon_circle-cross` to implement square text input

## Review points

<!-- List the points you would like to be reviewed in detail and the points you are not confident about. -->
<!-- Delete this section if not needed -->

- The number of file changes seems big, but most of them are because the directory structure has changed. So please focus the review on `RenameGroupForm`. 

## Screen Captures

<!-- Add screenshots or screen recordings of the changes -->
<!-- Necessary for UI changes -->
<!-- Delete this section if not needed -->

https://github.com/nishiki-tech/nishiki-frontend/assets/99339182/4db5dfca-9125-47ce-8794-2db764c87b88





## Assignee Checklist:

<!-- You can tick the checkboxes if not applicable -->

- [x] The naming convention of the PR title is correct (See the comment at the top of this template)
- [x] The base branch is correct (See: [Types of Branches](https://github.com/nishiki-tech/nishiki-frontend/blob/develop/docs/CONTRIBUTING.md#types-of-branches))
- [x] The branch name follows the [Branch Naming Conventions](https://github.com/nishiki-tech/nishiki-frontend/blob/develop/docs/CONTRIBUTING.md#branch-naming-conventions)
- [x] The correct assignees and reviewers have been designated for this PR
- [x] The coding style follows the [Coding Style Guide](https://github.com/nishiki-tech/nishiki-frontend/blob/develop/docs/STYLEGUIDE.md)
- [x] All the related issues are associated with this PR
- [x] All criteria in the associated issues are met (please tick the checkboxes)
- [x] My changes do not generate new warnings or errors (especially in the console)

## Reviewer Checklist:

<!-- You can tick the checkboxes if not applicable -->

- [x] Double-check the "Assignee Checklist"
- [x] The code follows the generic best practices and our [Coding Style Guide](https://github.com/nishiki-tech/nishiki-frontend/blob/develop/docs/STYLEGUIDE.md)
- [x] The code is well-commented and easy to understand
- [x] The UI changes accurately reflect the provided design

kanta1207 commented on src/features/groups/components/GroupCardList/GroupCard/GroupCardContent/ContainerCount.tsx:
File name changed just because the directory structure has changed.

kanta1207 commented on src/features/groups/components/GroupCardList/GroupCard/GroupCardContent/UserCount.tsx:
File name changed just because the directory structure has changed.

kanta1207 commented on src/features/groups/components/GroupCardList/GroupCard/GroupCardContent/index.tsx:
Separated this component from `GroupCard`, because we needed to use client component.

kanta1207 commented on src/features/groups/components/GroupCardList/GroupCard/GroupCardContent/GroupCardMenuButton/index.tsx:
Just added function to open the RenameForm

kanta1207 commented on src/features/groups/components/GroupCardList/GroupCard/index.tsx:
Separated some part of this component and created `GroupCardContent`, because we needed to use client component.

kanta1207 commented on src/features/groups/components/GroupCardList/GroupCard/GroupCardContent/RenameGroupForm.tsx:
Although I tried to write comments as much as I could, logic in this component might confuse you...
Please ask me anything, and I'm also open for verbal communication!

nick-y-ito replied:
Thank you for your hard work and great implementation! I want to let you know that I have tested it and it works perfectly! 

However, I noticed that the part handling the blur event might be overcomplicated. Although I initially thought that handling the `onblur` event was the best approach, it seems that detecting a click outside of the input could be a simpler solution, as you already know that the cross button is disturbing things. Please take a look at this link to get an idea:
https://chat.openai.com/share/6c5cfec0-e801-4ab9-87e0-28810deebeb1.

You can create the logic as a custom hook that detects a click outside.

Also, I think it would be better if the functionality of the cross button was implemented within the `<Input>` component. This way, it won't be exposed to the outside.

nick-y-ito commented on src/features/groups/components/GroupCardList/GroupCard/GroupCardContent/RenameGroupForm.tsx:
```diff
@@ -0,0 +1,172 @@
+  /**
+   * Focus the input when the form is opened.
+   */
+  useEffect(() => {
```
Whenever you face a problem caused by Radix, the first thing to do to resolve it is to explore its official document.
In there, you will find the `onCloseAutoFocus` and notice the right solution.

See: https://www.radix-ui.com/primitives/docs/components/dropdown-menu#content

nick-y-ito replied:
In our verbal discussion, we decided to not use `onCloseAutoFocus` as it affects other cases, such as closing the dropdown menu by pressing the [ESK] key, which is a side effect.

kanta1207 replied:
Thank you for researching!
After applied UI from wireframe correctly, this problem went away because the menu button is not displayed when form is open.

nick-y-ito reviewed (changes requested):
@kanta1207 
Thank you for your excellent work! It was so hard. It also took me quite long to understand the code and underlying issues.

I left several comments on the work. While some of them may be minor and can be dismissed, others may be quite important and require your attention. Additionally, I wanted to bring your attention to a UI problem that I have stated below.

 - [ ] Rename input does not accurately reflect the given design. Please take a look at the images below.

**Current looking**
<img width="300" alt="image" src="https://github.com/nishiki-tech/nishiki-frontend/assets/99148565/177f1b54-35bd-46bb-85e7-742279c07aff">

**Given design**
<img width="300" alt="image" src="https://github.com/nishiki-tech/nishiki-frontend/assets/99148565/e20ecc07-b15f-42dc-acc5-851b5f0e52fe">

nick-y-ito commented on src/features/groups/components/GroupCardList/GroupCard/GroupCardContent/RenameGroupForm.tsx:
```diff
@@ -0,0 +1,172 @@
+   * Handle the escape key down event.
+   * @param event - The keyboard event
+   */
+  const handleEscKeyDown = (event: KeyboardEvent<HTMLInputElement>) => {
```
On second thought, it may be better to exit the form without updating the name. I am sorry for the confusion.

kanta1207 commented on src/features/groups/components/GroupCardList/GroupCard/GroupCardContent/RenameGroupForm.tsx:
```diff
@@ -0,0 +1,172 @@
+                      type="text"
+                      variant="square"
+                      className={cn(
+                        'text-lg pr-10',
```
This is for preventing the text overlapping the cross button to happen.
https://github.com/nishiki-tech/nishiki-frontend/assets/99339182/4c89f4ac-60be-4d74-8acd-2deeed5400c3

Set it `10` because the width of cross button is `w-10`!

kanta1207 commented:
> @kanta1207 Thank you for your excellent work! It was so hard. It also took me quite long to understand the code and underlying issues.
> 
> I left several comments on the work. While some of them may be minor and can be dismissed, others may be quite important and require your attention. Additionally, I wanted to bring your attention to a UI problem that I have stated below.
> 
> * [ ]  Rename input does not accurately reflect the given design. Please take a look at the images below.
> 
> **Current looking** <img alt="image" width="300" src="https://private-user-images.githubusercontent.com/99148565/303896378-177f1b54-35bd-46bb-85e7-742279c07aff.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDc3NTQ2MzgsIm5iZiI6MTcwNzc1NDMzOCwicGF0aCI6Ii85OTE0ODU2NS8zMDM4OTYzNzgtMTc3ZjFiNTQtMzViZC00NmJiLTg1ZTctNzQyMjc5YzA3YWZmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjEyVDE2MTIxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ0ZWE5YzE2OTA4MjM4MjE1ZDI4ZWRhMTdiZjBhYzc4ZjViZDQ3OGU3YTliMTM0MzdhZDQwYTU3NWFiYjY2OTEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.H4P-GcBeEtCXwwImiNHMTlBWKYgLZ7neKPBStdQ4_E8">
> 
> **Given design** <img alt="image" width="300" src="https://private-user-images.githubusercontent.com/99148565/303896392-e20ecc07-b15f-42dc-acc5-851b5f0e52fe.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDc3NTQ2MzgsIm5iZiI6MTcwNzc1NDMzOCwicGF0aCI6Ii85OTE0ODU2NS8zMDM4OTYzOTItZTIwZWNjMDctYjE1Zi00MmRjLWFjYzUtODUxYjVmMGU1MmZlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjEyVDE2MTIxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM4YWFiYTVmNjlmYmZmNzIxNTQwMWVhZjFlOTA4NDQ3MDBmYTRjMTEyNTgyYTRjMmViMmU1MDgxNmY3NzMyZmImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.y6bZrBZFbdElglSpYbXT2GRxwJ0W5iu-oHBPG2MFUNM">

@nick-y-ito Thank you for the review! 
Following your review, I addressed the issues and reacted with 🚀 on the ones I resolved!
I'll leave some comments on the files I changed for the 2nd round review, so please refer them and take a look at the PR again 🙏

kanta1207 commented on src/components/ui/Input/SquareTextInput.tsx:
```diff
@@ -0,0 +1,96 @@
+  /**
+   * The height of the input. Default is 'sm'
+   */
+  h?: VariantProps<typeof inputVariants>['h'];
```
Added `h` prop, because it'll be change in desktop design

kanta1207 commented on src/components/ui/Input/SquareTextInput.tsx:
Decided to separate `SquareTextInput` component, basically for following reasons
- I felt it's too many logic to contain in base `Input` component, all of them won't be used for rounded input.
- To contain the `useOutsideClick` logic you shared with me in this file, instead of create it as. custom hook. The `useOutsideClick` logic is only used for `SquareTextInput` for now, so I thought it's better to contain it in here

nick-y-ito replied:
As we talked earlier verbally, let's take it out as a custom hook since it's a generic logic.

kanta1207 commented on src/components/ui/Input/SquareTextInput.tsx:
```diff
@@ -0,0 +1,96 @@
+          variant="none"
+          className={cn(
+            'bg-gray-lightest border-b border-primary py-2 pl-2 pr-10 text-lg',
+            className,
```
Defining `square` text input styles in here. 
`pr-10` is for preventing text from overlapping the cross button
Without it, it'll be like this
https://github.com/nishiki-tech/nishiki-frontend/assets/99339182/365218cf-8b3e-453d-9929-8dcd3453abc0

nick-y-ito replied:
I understood! Thank you for your explanation :)

kanta1207 commented on src/components/ui/Input/SquareTextInput.tsx:
```diff
@@ -0,0 +1,99 @@
+  /**
+   * Executed when outside of the input is clicked. Will not be invoked when cross button is clicked.
+   */
+  handleOutsideClick?: () => void;
```
handleOutsideClick is optional in here, since I wasn't sure if it's needed in everywhere this component is used.

kanta1207 commented on src/components/ui/Input/InputUsageExample.tsx:
Updated usage example file to apply `SquareTextInput` component

nick-y-ito commented:
@kanta1207 
Thank you for your excellent update! Overall, it works so well😃
As we talked earlier verbally, let's tackle the remaining two problems.

- [ ] The issue that pressing [ enter ] triggers the cross button click
- [ ] Refactoring the outside click logic creating a custom hook (https://github.com/nishiki-tech/nishiki-frontend/pull/223#discussion_r1488202117)

Other than those,  I believe that this pull request will be perfect and ready to be merged. Thanks again for all your hard work!

kanta1207 commented:
@nick-y-ito Thank you for your review!
I fixed and refactored the issues we discussed! 
> * [ ]  The issue that pressing [ enter ] triggers the cross button click
> * [ ]  Refactoring the outside click logic creating a custom hook ([Feat: Create `RenameGroupForm`  #223 (comment)]

I will leave some comments on the code as well, so please kindly take a look again 🙏

kanta1207 commented on src/components/ui/Input/SquareTextInput.tsx:
```diff
@@ -0,0 +1,101 @@
+  /**
+   * Function to clear the input value with cross button
+   */
+  handleClearInput?: () => void;
```
handleClearInput is optional in here, and if it's not provided, cross button will not be displayed.

kanta1207 commented on src/components/ui/Input/SquareTextInput.tsx:
```diff
@@ -0,0 +1,101 @@
+          variant="none"
+          className={cn(
+            'bg-gray-lightest border-b border-primary py-2 text-lg',
+            hasCrossButton ? 'pl-2 pr-10' : 'px-2',
```
Only when cross button is displayed, pr-10 is needed

nick-y-ito commented on src/features/groups/components/GroupCardList/GroupCard/GroupCardContent/RenameGroupForm.tsx:
```diff
@@ -0,0 +1,131 @@
+  useEffect(() => {
+    if (isOpen) {
+      inputRef.current?.focus();
+    }
```
This if statement can be written in one line

nick-y-ito reviewed (approved):
@kanta1207 
Thank you so much for your excellent update! I've just left some comments on tiny matters; if you'd like to address them, please feel free to do so. However, this PR already looks absolutely good to go 😄

kanta1207 commented:
> @kanta1207 Thank you so much for your excellent update! I've just left some comments on tiny matters; if you'd like to address them, please feel free to do so. However, this PR already looks absolutely good to go 😄

Thank you for your detailed review throughout this PR! I addressed the issues you mentioned, and I'll merge this PR now~ 👍
